src/Symfony/Component/HttpClient/Response/HttplugPromise.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Response;
+
+use GuzzleHttp\Promise\PromiseInterface;
+use Http\Client\Exception\NetworkException;
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface as Psr7ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @internal
+ */
+final class HttplugPromise implements Promise
+{
+    private $request;
+    private $response;
+    private $client;
+    private $responseFactory;
+    private $streamFactory;
+    private $result;
+    private $state = Promise::PENDING;
+    private $callbacks = [];
+
+    private static $promisePool;
+
+    /**
+     * @param ResponseInterface|\Exception $response
+     */
+    public function __construct(RequestInterface $request, $response, HttpClientInterface $client, ResponseFactoryInterface $responseFactory, StreamFactoryInterface $streamFactory)
+    {
+        if ($response instanceof \Exception) {
+            $this->reject($response);
+
+            return;
+        }
+
+        if (!$response instanceof ResponseInterface) {
+            throw new \TypeError(sprintf('Argument 2 passed to %s() must be an instance of %s|Exception, %s given.', __METHOD__, ResponseInterface::class, \gettype($response)));
+        }
+
+        $this->request = $request;
+        $this->response = $response;
+        $this->client = $client;
+        $this->responseFactory = $responseFactory;
+        $this->streamFactory = $streamFactory;
+
+        self::$promisePool = self::$promisePool ?? new \SplObjectStorage();
+        $clientPool = self::$promisePool[$client] ?? self::$promisePool[$client] = new \SplObjectStorage();
+        $clientPool[$response] = $this;
+    }
+
+    public function cancel(): void
+    {
+        if (Promise::PENDING !== $this->state) {
+            return;
+        }
+
+        $this->callbacks = [];
+
+        if ($this->response) {
+            $this->response->cancel();
+        } elseif ($this->result instanceof self || $this->result instanceof PromiseInterface) {
+            $this->result->cancel();
+        }
+
+        $this->wait(false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function then(callable $onFulfilled = null, callable $onRejected = null): Promise
+    {
+        if ($onFulfilled || $onRejected) {
+            $this->callbacks[] = [$onFulfilled, $onRejected];
```

Comment:
What if the promise is already fulfilled when you call “then”?

Reply:
Then nothing should happen: https://promisesaplus.com/ is explicit about that: the callbacks must be called on "next-tick". For us, that means on a subsequent call to `wait()`.

Reply:
Hmm but the corresponding logic in `wait()` is missing you're right.

Reply:
fixed